### PR TITLE
Storage persistence refactoring + `LastModified`.

### PR DESCRIPTION
**Pull request by mzhurovich**
- `meta.timestamp` -> `meta.begin_us` + `meta.end_us`.
- Added `.us` to all mutation types.
- Added `LastModified` to all containers.

### CODE REVIEW

File: Storage/base.h [hunk 1]
```diff
@@ -184,13 +190,13 @@ struct MutationJournal {
   }
 
   void Clear() {
-    meta_fields.clear();
+    transaction_meta.fields.clear();
     commit_log.clear();
     rollback_log.clear();
```
Comment (dkorolev) [inline Storage/base.h]: Perhaps nullify `begin_us` and `end_us` in `Clear()` and test it in `AssertEmpty()` as well?

File: Storage/container/dictionary.h [hunk 2]
```diff
@@ -58,31 +58,76 @@ class GenericDictionary {
...
-      journal_.LogMutation(UPDATE_EVENT(object),
-                           [this, key, previous_object]() { map_[key] = previous_object; });
+    const auto map_iterator = map_.find(key);
+    const auto lm_iterator = last_modified_.find(key);
+    if (map_iterator != map_.end()) {
+      const T& previous_object = map_iterator->second;
+      assert(lm_iterator != last_modified_.end());
+      const auto previous_timestamp = lm_iterator->second;
+      journal_.LogMutation(UPDATE_EVENT(now, object),
+                           [this, key, previous_object, previous_timestamp]() {
```
Comment (dkorolev) [inline Storage/container/dictionary.h]: Could you please confirm for me this `[..., previous_object, ...]` is guaranteed to capture by value, even though `previous_object` is declared as a const reference?

File: Storage/container/many_to_many.h [hunk 3]
```diff
@@ -105,12 +123,27 @@ class GenericManyToMany {
     return operator[](std::make_pair(row, col));
   }
 
+  ImmutableOptional<std::chrono::microseconds> LastModified(const key_t& key) const {
```
Comment (dkorolev) [inline Storage/container/many_to_many.h]: `sfinae::CF<key_t>`?
Or perhaps even `using`-declare it as `cf_key_t`?

**Comment by dkorolev** [comment]
СС @grixa -- FYI, we're adding `LastModified` to the all Storage fields.

File: Storage/container/one_to_many.h [hunk 4]
```diff
@@ -63,47 +64,82 @@ class GenericOneToMany {
...
-                           [this, key, previous_object]() { DoAdd(key, previous_object); });
+    const auto map_cit = map_.find(key);
+    const auto lm_cit = last_modified_.find(key);
+    if (map_cit != map_.end()) {
+      const T& previous_object = *(map_cit->second);
+      assert(lm_cit != last_modified_.end());
+      const auto previous_timestamp = lm_cit->second;
+      journal_.LogMutation(UPDATE_EVENT(now, object),
+                           [this, key, previous_object, previous_timestamp]() {
+                             DoAddWithLastModified(previous_timestamp, key, previous_object);
```
Comment (dkorolev) [inline Storage/container/one_to_many.h]: Here and above, `DoUpdateWithLastModified` is probably a better name -- it's in fact reverting the record, not adding it.

File: Storage/container/many_to_many.h (hunk 3 above)
Comment (mzhurovich) [inline Storage/container/many_to_many.h]: `key_t` is always an `std::pair` for matrix containers.

Comment (dkorolev) [inline Storage/container/many_to_many.h]: :blush:
A comment would be nice.

File: Storage/container/one_to_many.h (hunk 4 above)
Comment (mzhurovich) [inline Storage/container/one_to_many.h]: ???
It adds the record and sets the modification timestamp, nothing more.

Comment (dkorolev) [inline Storage/container/one_to_many.h]: Except "adds" may be "overwrites with the older version", and "sets timestamp" may be "drags the timestamp into the past".

File: Storage/container/many_to_many.h (hunk 3 above)
Comment (mzhurovich) [inline Storage/container/many_to_many.h]: Where do you want to see it?
It's self-explanatory by definition :)
`using key_t = std::pair<row_t, col_t>;`

Comment (dkorolev) [inline Storage/container/many_to_many.h]: Above the very method, along the lines of `Pass the key by a const reference, as `key_t` is an `std::pair<row_t, col_t>`..

File: Storage/container/one_to_one.h [hunk 5]
```diff
@@ -189,22 +255,29 @@ class GenericOneToOne {
...
+    forward_[key.first] = placeholder.get();
+    transposed_[key.second] = placeholder.get();
+    last_modified_[key] = us;
+  }
+
   void DoErase(const key_t& key) {
     forward_.erase(key.first);
     transposed_.erase(key.second);
     map_.erase(key);
   }
```
Comment (dkorolev) [inline Storage/container/one_to_one.h]: + `last_modified_.erase(key)`.

Comment (dkorolev) [inline Storage/container/one_to_one.h]: Or, assuming you thought of it more than I did and my comment is misleading, rename `DoErase` into some `DoEraseFromMap` / `DoEraseWithoutTouchingLastModified`, etc.

File: Storage/container/one_to_one.h [hunk 6]
```diff
@@ -189,22 +255,29 @@ class GenericOneToOne {
...
+    DoErase(key);
+    last_modified_[key] = us;
   }
 
   elements_map_t map_;
   forward_map_t forward_;
   transposed_map_t transposed_;
+  std::unordered_map<key_t, std::chrono::microseconds, CurrentHashFunction<key_t>> last_modified_;
   MutationJournal& journal_;
 };
```
Comment (dkorolev) [inline Storage/container/one_to_one.h]: Note to self: Reviewed until here for now.

File: Storage/container/dictionary.h (hunk 2 above)
Comment (mzhurovich) [inline Storage/container/dictionary.h]: I swear :)
> The type of each data member is the type of the corresponding captured entity, except if the entity has reference type (in that case, references to functions are captured as lvalue references to the referenced functions, and references to objects are captured as copies of the referenced objects).

Comment (dkorolev) [inline Storage/container/dictionary.h]: Равнобедренный! :)

File: Storage/docu/docu_2_code.cc [hunk 7]
```diff
@@ -206,41 +210,58 @@ TEST(StorageDocumentation, BasicUsage) {
...
+
+    EXPECT_EQ(1000, t.meta.begin_us.count());
+    EXPECT_EQ(1003, t.meta.end_us.count());
     EXPECT_EQ(1u, t.meta.fields.size());
     EXPECT_EQ("max", t.meta.fields.at("user"));
   }
 
   {
-    const auto t = ParseAndValidateRow(persisted_transactions[1], 1u, std::chrono::microseconds(1004));
+    const auto t = ParseAndValidateRow(persisted_transactions[1], 1u, std::chrono::microseconds(1102));
```
Comment (dkorolev) [inline Storage/docu/docu_2_code.cc]: :black_circle: magic.

File: Storage/golden/transactions_to_replicate.json [hunk 8]
```diff
@@ -1,2 +1,2 @@
-{"index":0,"us":100}	{"meta":{"timestamp":100,"fields":{"user":"dima"}},"mutations":[{"RecordDictionaryUpdated":{"data":{"lhs":"one","rhs":1}},"":"T9205381019427680739"},{"RecordDictionaryUpdated":{"data":{"lhs":"two","rhs":2}},"":"T9205381019427680739"}]}
-{"index":1,"us":200}	{"meta":{"timestamp":200,"fields":{}},"mutations":[{"RecordDictionaryUpdated":{"data":{"lhs":"three","rhs":3}},"":"T9205381019427680739"},{"RecordDictionaryDeleted":{"key":"two"},"":"T9202204057466171294"}]}
+{"index":0,"us":103}	{"meta":{"begin_us":100,"end_us":103,"fields":{"user":"dima"}},"mutations":[{"RecordDictionaryUpdated":{"us":101,"data":{"lhs":"one","rhs":1}},"":"T9200018162904582576"},{"RecordDictionaryUpdated":{"us":102,"data":{"lhs":"two","rhs":2}},"":"T9200018162904582576"}]}
+{"index":1,"us":203}	{"meta":{"begin_us":200,"end_us":203,"fields":{}},"mutations":[{"RecordDictionaryUpdated":{"us":201,"data":{"lhs":"three","rhs":3}},"":"T9200018162904582576"},{"RecordDictionaryDeleted":{"us":202,"key":"two"},"":"T9200749443989249031"}]}
```
Comment (dkorolev) [inline Storage/golden/transactions_to_replicate.json]: :+1:

File: Storage/test.cc [hunk 9]
```diff
@@ -1257,14 +1257,348 @@ TEST(TransactionalStorage, TransactionMetaFields) {
...
+        ASSERT_TRUE(Exists(t));
+        EXPECT_EQ(312, Value(t).count());
+      }
+      current::time::SetNow(std::chrono::microseconds(313));
+      fields.uone_to_uone.Erase(1, "y");
+      {
+        const auto t = fields.uone_to_uone.LastModified(1, "y");
+        ASSERT_TRUE(Exists(t));
+        EXPECT_EQ(313, Value(t).count());
+      }
```
Comment (dkorolev) [inline Storage/test.cc]: Me wow.

**Comment by dkorolev** [comment]
LGTM.

My only concern is that once a user-initiated modification triggers one or two more (due to conflicts on `one_to_many` and `one_to_one`), why not call `Now()` multiple times to ensure individual data field mutations have unique timestamps?

Thanks,
Dima

File: Storage/container/one_to_one.h (hunk 6 above)
Comment (dkorolev) [inline Storage/container/one_to_one.h]: Resolved, reviewed all.

File: Storage/container/many_to_many.h (hunk 3 above)
Comment (mzhurovich) [inline Storage/container/many_to_many.h]: 👍

File: Storage/base.h (hunk 1 above)
Comment (mzhurovich) [inline Storage/base.h]: 👍

File: Storage/container/one_to_one.h (hunk 5 above)
Comment (mzhurovich) [inline Storage/container/one_to_one.h]: 👍

**Comment by mzhurovich** [comment]
@dkorolev PTAL

**Pull request merged**